File: algorithms/timus-online-judge/graphs/1527_BadRoads.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

struct Edge {
  int vtxFrom, vtxTo, cost, time, height;

  Edge(int vtxFrom, int vtxTo, int cost, int time, int height)
  : vtxFrom(vtxFrom), vtxTo(vtxTo), cost(cost), time(time), height(height)
  { }
};

struct Constraints {
  int totalCost, totalTime, maxCost, maxTime, maxHeight;

  Constraints(int totalCost, int totalTime, int maxCost, int maxTime, int maxHeight)
    : totalCost(totalCost)
    , totalTime(totalTime)
    , maxCost(maxCost)
    , maxTime(maxTime)
    , maxHeight(maxHeight)
  { }
};

const int INF = 1000000000;

std::vector<int> findPath(
  int numVertices,
  int source, int target,
  const std::vector<Edge> &edges,
  const Constraints &constraints
) {
  int times[numVertices][numVertices];
  int paths[numVertices][numVertices];
  std::vector<int> path;

  for (int i = 0; i < numVertices; i++) {
    for (int j = 0; j < numVertices; j++) {
      times[i][j] = INF;
      paths[i][j] = -1;
    }
  }

  times[source][0] = 0;
  for (int currCost = 0; currCost < numVertices; currCost++) {
    while (true) {
      bool updated = false;

      for (int edgeId = 0; edgeId < edges.size(); edgeId++) {
        Edge edge = edges[edgeId];

        if (edge.height > constraints.maxHeight) {
          continue;
        }

        int updatedTime = INF;

        if (edge.cost == 0) {
          updatedTime = times[edge.vtxFrom][currCost] + edge.time;
        } else if (currCost > 0) {
          updatedTime = times[edge.vtxFrom][currCost - 1] + edge.time;
        }
        
        if (updatedTime < times[edge.vtxTo][currCost]) {
          times[edge.vtxTo][currCost] = updatedTime;
          paths[edge.vtxTo][currCost] = edgeId;
          updated = true;
        }

        if (edge.vtxTo == target &&
          currCost <= constraints.maxCost &&
          times[target][currCost] <= constraints.maxTime
        ) {
          while (edgeId != -1) {
            path.push_back(edgeId + 1);
            target = edges[edgeId].vtxFrom;
            currCost -= edges[edgeId].cost;
            edgeId = paths[target][currCost];            
          }

          return path;
        }
      }

      if (!updated) {
        break;
      }
    }
  }

  return path;
}

int main() {
  std::ios_base::sync_with_stdio(false);

  int numVertices, numEdges, source, target, maxCost, maxTime;
  int minHeight = INF;
  int maxHeight = 0;

  std::cin >> numVertices >> numEdges >> source >> target >> maxCost >> maxTime;

  source -= 1;
  target -= 1;
  
  std::vector<Edge> edges;
  int vtxFrom, vtxTo, cost, time, height;
  for (int i = 0; i < numEdges; i++) {
    std::cin >> vtxFrom >> vtxTo >> cost >> time >> height;
    edges.push_back(Edge(
      vtxFrom - 1, vtxTo - 1, cost, time, height
    ));
    minHeight = std::min(minHeight, height);
    maxHeight = std::max(maxHeight, height);
  }

  Constraints constraints(0, 0, maxCost, maxTime, maxHeight);

  std::vector<int> path = findPath(
    numVertices,
    source, target,
    edges, constraints
  );

  if (path.size() == 0) {
    std::cout << -1 << std::endl;
    return 0;
  }

  int left = minHeight;
  int right = maxHeight;
  int optimalHeight = maxHeight;

  while (left <= right) {
    int currHeight = (left + right) / 2;
    Constraints currConstraints(0, 0, maxCost, maxTime, currHeight);

    std::vector<int> currPath = findPath(
      numVertices,
      source, target,
      edges, currConstraints
    );
    
    if (currPath.size() == 0) {
      left = currHeight + 1;
    } else {
      path = std::move(currPath);
      optimalHeight = currHeight;
      right = currHeight - 1;
    }
  }

  std::cout << optimalHeight << std::endl;
  std::cout << path.size() << std::endl;
  for (int i = 0; i < path.size(); i++) {
    std::cout << path[path.size() - i - 1] << " ";
  }

  return 0;
}

File: algorithms/timus-online-judge/graphs/1527_BadRoads.py


class Edge:
    def __init__(self, vtx_from, vtx_to, cost, time, height):
        self.vtx_from = vtx_from
        self.vtx_to = vtx_to
        self.cost = cost
        self.time = time
        self.height = height


def find_paths(n, src, dst, edges, max_cost, max_time, max_height):
    times = [[float("inf") for _ in range(n)] for _ in range(n)]
    paths = [[-1 for _ in range(n)] for _ in range(n)]

    times[src][0] = 0
    for curr_cost in range(n):
        while True:
            updated = False

            for edge_id, edge in enumerate(edges):
                if edge.height > max_height:
                    continue

                updated_time = float("inf")

                if edge.cost == 0:
                    updated_time = times[edge.vtx_from][curr_cost] + edge.time
                elif curr_cost > 0:
                    updated_time = times[edge.vtx_from][curr_cost - 1] + edge.time
                
                if updated_time < times[edge.vtx_to][curr_cost]:
                    times[edge.vtx_to][curr_cost] = updated_time
                    paths[edge.vtx_to][curr_cost] = edge_id
                    updated = True

                if edge.vtx_to == dst and \
                   curr_cost <= max_cost and \
                   times[dst][curr_cost] <= max_time:
                    return curr_cost, paths
                
            if not updated:
                break
    
    return -1, []


def build_path(cost, edges, paths, dst):
    curr_vtx, curr_cost, path = dst, cost, []

    edge_id = paths[curr_vtx][curr_cost]
    while edge_id != -1:
        path.append(edge_id + 1)
        curr_vtx = edges[edge_id].vtx_from
        curr_cost -= edges[edge_id].cost
        edge_id = paths[curr_vtx][curr_cost]

    return path[::-1]


def main():
    n, m, src, dst = map(int, input().split())
    max_cost, max_time = map(int, input().split())

    src -= 1
    dst -= 1

    edges = []
    min_height, max_height = float("inf"), 0

    for _ in range(m):
        vtx_from, vtx_to, cost, time, height = map(int, input().split())
        edges.append(Edge(vtx_from - 1, vtx_to - 1, cost, time, height))
        min_height = min(min_height, height)
        max_height = max(max_height, height)
    
    cost, paths = find_paths(
        n, src, dst, edges,
        max_cost, max_time,
        max_height
    )

    if cost == -1:
        print(-1)
        return
    
    optimal_height = max_height
    left, right = min_height, max_height

    while left <= right:
        curr_height = (left + right) // 2

        curr_cost, curr_paths = find_paths(
            n, src, dst, edges,
            max_cost, max_time,
            curr_height
        )

        if curr_cost == -1:
            left = curr_height + 1
        else:
            cost = curr_cost
            paths = curr_paths
            optimal_height = curr_height
            right = curr_height - 1

    path = build_path(cost, edges, paths, dst)

    print(optimal_height)
    print(len(path))
    print(" ".join(map(str, path)))


if __name__ == "__main__":
    main()
